src/event_loop.rs

```diff
@@ -378,6 +262,54 @@ impl<T> EventLoopWindowTarget<T> {
         #[cfg(any(x11_platform, wasm_platform, wayland_platform, windows))]
         self.p.listen_device_events(_allowed);
     }
+
+    /// When the current loop iteration finishes, immediately begin a new iteration regardless of
+    /// whether or not new events are available to process.
+    ///
+    /// This is the default.
+    pub fn set_poll(&self) {
+        self.p.set_poll()
+    }
+
+    /// When the current loop iteration finishes, suspend the thread until another event arrives.
+    pub fn set_wait(&self) {
+        self.p.set_wait()
+    }
+
+    /// When the current loop iteration finishes, suspend the thread until either another event
+    /// arrives or the given time is reached.
+    ///
+    /// Useful for implementing efficient timers. Applications which want to render at the display's
+    /// native refresh rate should instead use [`set_poll()`] and the VSync functionality of a graphics API
+    /// to reduce odds of missed frames.
+    ///
+    /// [`set_poll()`]: Self::set_poll
+    pub fn set_wait_until(&self, instant: Instant) {
+        self.p.set_wait_until(instant)
+    }
+
+    /// Sets the event loop to wait until a timeout has expired.
+    ///
+    /// In most cases, this is uses [`set_wait_until()`]. However, if the timeout overflows, it
+    /// instead uses [`set_wait()`].
+    ///
+    /// [`set_wait_until()`]: Self::set_wait_until
+    /// [`set_wait()`]: Self::set_wait
+    pub fn set_wait_timeout(&self, timeout: Duration) {
+        match Instant::now().checked_add(timeout) {
+            Some(instant) => self.set_wait_until(instant),
+            None => self.set_wait(),
+        }
+    }
+
+    /// Send a [`LoopExiting`] event and stop the event loop. This variant is *sticky* - once set,
+    /// it cannot be changed, and any future attempts to do so will
+    /// result in it being reset to it's previous state.
+    ///
+    /// [`LoopExiting`]: Event::LoopExiting
+    pub fn exit(&self) {
```

Comment:
Also, given that users may want to propagate the error somehow, maybe we should allow them to pass `payload` with the `exit` somehow? Likely via the means of dynamic dispatch or straight a `Result<(), Box<dyn Error>>`? But it could be handled in a separate PR, it's just a pain point right now to propagate your error outside of the event loop `run` methods.

Reply:
Yeah I would prefer to figure that out in a separate PR, assuming the discussion in #3059 hasn't concluded the issue already.